portfolio/src/main/java/com/google/sps/servlets/CommentsDatastore.java

```diff
@@ -0,0 +1,70 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.util.List;
+import java.util.ArrayList;
+import com.google.gson.Gson;
+import com.google.appengine.api.datastore.*;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+public class CommentsDatastore {
+
+  public class Comment {
+    private String comment;
+    private long timestamp;
+
+    public Comment(String comment, long timestamp) {
+      this.comment = comment;
+      this.timestamp = timestamp;
+    }
+  } 
+
+    public CommentsDatastore() {}
+    
+    public List<Comment> fetchComments(int maxNumOfComments) {
+      Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      List<Entity> results = datastore.prepare(query)
+        .asList(FetchOptions.Builder.withLimit(maxNumOfComments));
+      
+      List<Comment> comments = new ArrayList<>();
+      for (Entity entity : results) {
+        String comment = (String) entity.getProperty("comment");
+        long timestamp = (long) entity.getProperty("timestamp");
+        Comment newComment = new Comment(comment, timestamp);
+        comments.add(newComment);
+      }
+
+      return comments;
+    }
+
+    public Comment addComment(String comment, int maxNumOfComments) {
+      long timestamp = System.currentTimeMillis();
```

Comment:
FYI (don't need to change this for this mini project, just for your learning), we rarely use System.currentTimeMillis like this because it doesn't work in tests as we cannot override the value so instead we usually create a Clock interface with getTime() and have a real clock returning time from System and a FakeClock in test that can override what we return.

Reply:
That makes sense. 
For the capstone project, will we need to write tests, or is it enough for us to test manually on the live server?

Reply:
capstone we will want tests. don't worry about it for your mini projects